### ISSUE TITLE
export/save models with a layer_lambda()

### ISSUE THREAD

**Issue by t-kalinowski** [rstudio/keras#274]
Hello and thanks for the great package!

Is it possible to call `export_savedmodel()` or `save_model_hdf5()` on a model that includes a `layer_lambda()`?

When I try to do that I get:
```
TypeError: can't pickle PyCapsule objects
```

I think this touched on in #90, where the suggested solution was to use `save_model_weights_hdf5()` and redefine the model in R each time instead. That is great during development, but won't work when the models need to run without any R or python runtime dependency. (e.g, with `tfdeploy`). 

With some googeling it seems it might be possible to save the tensorflow graph, doing something like
https://blog.keras.io/keras-as-a-simplified-interface-to-tensorflow-tutorial.html#exporting-a-model-with-tensorflow-serving
or 
https://github.com/amir-abdi/keras_to_tensorflow/blob/master/keras_to_tensorflow.py

I gave a quick try for that strategy but wasn't immediately successful. Before I keep on this path I thought I'd file this issue and ask:
a) Is this even possible?
b) Are there plans to support these type of models in `export_savedmodel()`?
c) Do you have suggestions about what the best (most reliable, most robust) strategy here might be?

**Comment by jjallaire** [comment]
I would have indeed suggested saving the model weights but as you point out that can be problematic for deployment. We do have a Keras implementation of `export_savedmodel()` in the most recent CRAN version of the Keras package which works around some of the issues you might have seen. What happens if you use `keras::export_savedmodel()`?

**Comment by t-kalinowski** [comment]
with CRAN keras and tensorflow:
```
> export_savedmodel(model)
Keras learning phase set to 0 for export (restart R session before doing additional training)

 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  TypeError: can't pickle PyCapsule objects 
```

If it makes any difference, my `layer_lambda` is relatively simple and uses tensorflow ops
```
  layer_lambda(function(x) {
      cx <- tf$complex(real = x[, , 1], imag = x[, , 2])
      ft <- tf$fft(cx)
      tf$stack(list(tf$real(ft), tf$imag(ft),
                    tf$angle(ft), tf$abs(ft)), axis = -1)
    })
```

**Comment by javierluraschi** [comment]
@t-kalinowski any chance you could share all the code you are using to reproduce this?

**Comment by t-kalinowski** [comment]
Sure,
I think that a truly minimal example would just be a `keras_sequental_model` that consists of just one `layer_lambda`, however, here is a larger example that takes advantage of the, potentially complicating, features of the functional API.

``` R
library(tensorflow)
library(keras)

input <- layer_input(shape = c(128, 2))

time_domain_view <- input %>%
  layer_conv_1d(128, kernel_size = 7, activation = "relu") %>%
  layer_max_pooling_1d(5) %>%
  layer_flatten() %>%
  layer_dense(128, activation = 'relu')

freq_domain_view <- input %>%
  layer_lambda(function(x) {
    cx <- tf$complex(real = x[, , 1], imag = x[, , 2])
    ft <- tf$fft(cx)
    tf$stack(list(tf$real(ft), tf$imag(ft),
                  tf$angle(ft), tf$abs(ft)), axis = -1)
  }) %>%
  layer_conv_1d(128, kernel_size = 7, activation = 'relu') %>%
  layer_max_pooling_1d(pool_size = 5) %>%
  layer_flatten() %>%
  layer_dense(64, activation = 'relu')

combined <-
  layer_concatenate(list(time_domain_view, freq_domain_view))

classification <- combined %>%
  layer_dense(10, activation = 'softmax')

model <- keras_model(input, classification)

model %>%
  compile(loss = 'categorical_crossentropy',
          optimizer = 'adam',
          metrics = 'accuracy')
```

```
> export_savedmodel(model)
Keras learning phase set to 0 for export (restart R session before doing additional training)

 Error in py_call_impl(callable, dots$args, dots$keywords) : 
  TypeError: can't pickle PyCapsule objects 
```

**Comment by javierluraschi** [comment]
@t-kalinowski Currently, keras removes the learning phase tensor that is introduced by some keras layers, the `layer_lambda` causes this operation to fail. We have a fix ready for you here https://github.com/rstudio/keras/pull/275 that can be installed as follows:

If you can't wait to try this you could run:

```r
devtools::install_github("rstudio/keras")
```

then you will have to use this for the export:

```
export_savedmodel(model, "savedmodel", remove_learning_phase = FALSE)
```

**Comment by t-kalinowski** [comment]
Thank you for the quick resolution!

Would this solution still be appropriate for me if I had, for example, a `layer_dropout()` in my model? 
In that case, would I be responsible for setting `k_set_learning_phase(0)` before defining the model on my own?

In that case, would something like this be sufficient? 
```
# in a fresh R session
library(keras)

k_set_learning_phase(0)

source("define_my_model.R")  # script defines object "model" in the global env
# "model" contains both a layer_lambda() and a layer_dropout()

model %>% load_model_weights_hdf5("my_trained_model_weights.h5")

model %>% export_savedmodel("my_exported_model", remove_learning_phase = FALSE)
```

**Comment by javierluraschi** [comment]
@t-kalinowski I believe `k_set_learning_phase(0)` would have to be set before training for the export to not contain the learning phase; however, this is not a good approach since `k_set_learning_phase(0)` would affect training.

There seems to be no way to remove the learning phase tensor unless we do what the Keras blog recommends: https://blog.keras.io/keras-as-a-simplified-interface-to-tensorflow-tutorial.html

However, this workaround is the one provided in `export_savedmodel()` already and is the workaround that caused this issue in the first place.

Therefore, if you had a lambda layer and dropout, I would still use `remove_learning_phase = FALSE` which would work just fine but would have an additional input tensor that you would have to set while performing predictions. I don't think this is terrible, but something to be aware of.

**Comment by t-kalinowski** [comment]
Sorry, I'm not sure I'm understanding. Let me rephrase my current understanding how I should approach this:

Let's say my model has been trained in a separate session, and the desired final model weights have been saved to  `"my_trained_model_weights.h5"`. 

To get ready for export, I would start a fresh R session where no further training will be performed.
Then in that fresh R (and python) session I would redefine my model architecture like this:
```
k_set_learning_phase(0)
model <- keras_model_sequential() %>%
  ... %>%
  layer_lambda(...) %>%
  ...
  layer_dropout() %>%
  ...
  compile(...)
``` 
load the pretrained weights for that model like this:
```
model %>% load_model_weights_hdf5("my_trained_model_weights.h5")
```
and then export the model like this:
```
model %>% export_savedmodel("my_exported_model", remove_learning_phase = FALSE)
```

Is that sufficient, or am I missing a step?

**Comment by javierluraschi** [comment]
Sorry for the confusion, t-kalinowski. No, you don't want to set `k_set_learning_phase(0)` at all, after training and without restarting the session, you would want to run:

```r
model %>% export_savedmodel("my_exported_model", remove_learning_phase = FALSE)
```

That's it.

**Comment by t-kalinowski** [comment]
Ok, thank you. So if i'm understanding correctly, the exported model will still have `layer_dropout()` included in the tensorflow graph, but the `predict()` (or it's equivalent) on the deployed platform will be responsible for making sure that the dropout is inactive during predictions. Is that correct?

**Comment by javierluraschi** [comment]
@t-kalinowski that is correct!

**Comment by t-kalinowski** [comment]
Thank you very much!

**Comment by nandofernandesneto** [comment]
Btw, how you guys load a model after doing:

`model %>% export_savedmodel("my_exported_model", remove_learning_phase = FALSE)`

Which command do you use to load them afterwards?

Best,

Fernando
